### PR TITLE
Using ImmutableDruidDataSource as a key for map and set instead of DruidDataSource

### PR DESCRIPTION
**Pull request by jihoonson**
DruidDataSource is mutable and thus should not be used as a key of map or set.

---
This change is [Reviewable](https://reviewable.io/reviews/druid-io/druid/5054)

### CODE REVIEW

File: server/src/main/java/io/druid/client/DruidDataSource.java [hunk 1]
```diff
@@ -49,8 +49,8 @@ public DruidDataSource(
     this.name = name;
     this.properties = properties;
 
-    this.partitionNames = Maps.newHashMap();
-    this.segmentsHolder = new ConcurrentSkipListSet<DataSegment>();
+    this.idToSegmentMap = new HashMap<>();
+    this.segmentsHolder = new HashSet<>();
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/DruidDataSource.java]: is concurrency not needed?

File: server/src/main/java/io/druid/client/DruidDataSource.java [hunk 2]
```diff
@@ -38,8 +38,8 @@
 
   private final String name;
   private final Map<String, String> properties;
-  private final Map<String, DataSegment> partitionNames;
-  private final ConcurrentSkipListSet<DataSegment> segmentsHolder;
+  private final Map<String, DataSegment> idToSegmentMap;
+  private final Set<DataSegment> segmentsHolder;
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/DruidDataSource.java]: At a cursory glance it doesn't even look like we need segmentsHolder. A `ConcurrentHashMap` for `idToSegmentMap` should suffice for all use cases. This would also remove the "Big Kernel Lock" in this class.

File: server/src/main/java/io/druid/client/DruidDataSource.java [hunk 3]
```diff
@@ -128,52 +126,20 @@ public ImmutableDruidDataSource toImmutableDruidDataSource()
...
+             '}';
     }
-
-    return true;
-  }
-
-  @Override
-  public int hashCode()
-  {
-    int result = name != null ? name.hashCode() : 0;
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/DruidDataSource.java]: (optional) an alternative could be to throw `UOE` to catch any bad uses of this class.

File: server/src/main/java/io/druid/client/DruidDataSource.java (hunk 2 above)
Comment (drcrallen) [reply server/src/main/java/io/druid/client/DruidDataSource.java]: I took an opportunity to give a sho at https://github.com/druid-io/druid/pull/5056 , which changes the concurrency pattern in DruidServer. I think it could be changed here as well.

File: server/src/main/java/io/druid/client/HttpServerInventoryView.java [hunk 4]
```diff
@@ -294,18 +295,11 @@ public DruidServer getInventoryValue(String containerKey)
...
-        servers.values(), new Function<DruidServerHolder, DruidServer>()
-        {
-          @Override
-          public DruidServer apply(DruidServerHolder input)
-          {
-            return input.druidServer;
-          }
-        }
-    );
+    return servers.values().stream()
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/HttpServerInventoryView.java]: I *think* this change from lazy to eager allocation, but I don't think it will be an issue here.

Comment (drcrallen) [reply server/src/main/java/io/druid/client/HttpServerInventoryView.java]: A lot of calls actually use iterations over the inventory. In the near future it might make sense to have this method return a `Stream` so the caller can filter as needed.

File: server/src/main/java/io/druid/client/ImmutableDruidDataSource.java [hunk 5]
```diff
@@ -73,14 +69,52 @@ public boolean isEmpty()
...
+    // idToSegments is intentionally ignored because it is usually large
     return "ImmutableDruidDataSource{"
            + "name='" + name
            + "', segments='" + segmentsHolder
            + "', properties='" + properties
            + "'}";
   }
+
+  @Override
+  public boolean equals(Object o)
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/ImmutableDruidDataSource.java]: in a future PR this is changed to just compare the metadata. I was having trouble getting some tests to pass otherwise.

Comment (drcrallen) [reply server/src/main/java/io/druid/client/ImmutableDruidDataSource.java]: nothing need be done here.

File: server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorLogger.java [hunk 6]
```diff
@@ -259,7 +259,7 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
     final Stream<DataSegment> allSegments = params
         .getDataSources()
         .stream()
-        .flatMap((final DruidDataSource dataSource) -> dataSource.getSegments().stream());
+        .flatMap((ImmutableDruidDataSource dataSource) -> dataSource.getSegments().stream());
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorLogger.java]: is the type even needed here?

Comment (drcrallen) [reply server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorLogger.java]: IMHO this can be a bit cleaner with something like 

```
params
  .getDataSources()
  .stream()
  .map(DruidDataSource::getSegments)
  .flatMap(Set::stream);
```

File: server/src/main/java/io/druid/server/http/InventoryViewUtils.java [hunk 7]
```diff
@@ -37,42 +33,22 @@
...
-                      {
-                        return input.getDataSources();
-                      }
-                    }
-                )
-            )
-        )
-    );
-    return dataSources;
+    return serverInventoryView.getInventory().stream()
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: Suggest putting `.getInventory()` and `.stream()` on their own lines

File: server/src/main/java/io/druid/server/http/InventoryViewUtils.java [hunk 8]
```diff
@@ -37,42 +33,22 @@
...
+                              .flatMap(server -> server.getDataSources().stream())
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: Same comment as before about suggesting a `.map` followed by a `flatMap`

File: server/src/main/java/io/druid/client/DruidDataSource.java (hunk 1 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/DruidDataSource.java]: Please see the below comment.

File: server/src/main/java/io/druid/client/DruidDataSource.java (hunk 2 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/DruidDataSource.java]: Good point. I removed segmentsHolder and changed idToSegmentMap to ConcurrentHashMap. Also, thanks for raising #5056.

File: server/src/main/java/io/druid/client/DruidDataSource.java (hunk 3 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/DruidDataSource.java]: Thanks. Added.

File: server/src/main/java/io/druid/client/HttpServerInventoryView.java (hunk 4 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/HttpServerInventoryView.java]: Yeah, I guess eager allocation is not an issue because the return value is iterated right away in all places.

File: server/src/main/java/io/druid/client/ImmutableDruidDataSource.java (hunk 5 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/ImmutableDruidDataSource.java]: I guess you may refer to ImmutableDruidServer? ImmutableDruidDataSource doesn't have metadata.

File: server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorLogger.java (hunk 6 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorLogger.java]: Removed.

File: server/src/main/java/io/druid/server/http/InventoryViewUtils.java (hunk 7 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: Moved.

File: server/src/main/java/io/druid/server/http/InventoryViewUtils.java (hunk 8 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: ```flatMap``` should be followed by ```map``` here because ```server.getDataSources()``` returns a ```Collection<DruidDataSource>``` which should be flatten before applying ```map```.

File: server/src/main/java/io/druid/client/DruidDataSource.java [hunk 9]
```diff
@@ -65,115 +57,66 @@ public String getName()
     return properties;
   }
 
-  @JsonProperty
-  public Set<DataSegment> getSegments()
+  public Collection<DataSegment> getSegments()
   {
-    return Collections.unmodifiableSet(segmentsHolder);
+    return idToSegmentMap.values();
```
Comment (leventov) [inline server/src/main/java/io/druid/client/DruidDataSource.java]: Can use `unmodifiableCollection()`

File: server/src/main/java/io/druid/client/DruidDataSource.java [hunk 10]
```diff
@@ -65,115 +57,66 @@ public String getName()
...
-      return false;
-    }
-    if (segmentsHolder != null ? !segmentsHolder.equals(that.segmentsHolder) : that.segmentsHolder != null) {
-      return false;
-    }
-
-    return true;
+    return "DruidDataSource{" +
+           "properties=" + properties +
+           ", partitions=" + getSegments().toString() +
```
Comment (leventov) [inline server/src/main/java/io/druid/client/DruidDataSource.java]: toString() not needed

File: server/src/main/java/io/druid/client/ImmutableDruidDataSource.java [hunk 11]
```diff
@@ -58,29 +55,58 @@ public String getName()
...
   {
-    return segmentsHolder;
+    return idToSegments.get(segmentIdentifier);
   }
 
   @Override
   public String toString()
   {
-    // partitionNames is intentionally ignored because it is usually large
+    // idToSegments is intentionally ignored because it is usually large
```
Comment (leventov) [inline server/src/main/java/io/druid/client/ImmutableDruidDataSource.java]: Now not ignored

File: server/src/main/java/io/druid/client/ImmutableDruidServer.java [hunk 12]
```diff
@@ -128,4 +128,29 @@ public String toString()
...
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    DruidServer that = (DruidServer) o;
+
+    if (getName() != null ? !getName().equals(that.getName()) : that.getName() != null) {
```
Comment (leventov) [inline server/src/main/java/io/druid/client/ImmutableDruidServer.java]: Please annotate `DruidServerMetadata.getName()` as `@Nullable`.

Please use `Objects.equals()`, or regenerate equals and hashCode using intelliJ

File: server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java [hunk 13]
```diff
@@ -163,19 +163,11 @@ public ContainerClass getInventoryValue(String containerKey)
...
-        new Function<ContainerHolder, ContainerClass>()
-        {
-          @Override
-          public ContainerClass apply(ContainerHolder input)
-          {
-            return input.getContainer();
-          }
-        }
-    );
+    return containers.values().stream()
```
Comment (leventov) [inline server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java]: Break line before `.stream()`

File: server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java [hunk 14]
```diff
@@ -460,21 +452,11 @@ public void moveSegment(
...
-            {
-              @Override
-              public Iterable<DataSegment> apply(DruidDataSource input)
-              {
-                return input.getSegments();
-              }
-            }
-        )
-    );
+    return metadataSegmentManager.getInventory().stream()
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java]: Same

File: server/src/main/java/io/druid/server/http/DatasourcesResource.java [hunk 15]
```diff
@@ -496,40 +493,32 @@ public Response getSegmentDataSourceTiers(
     return Response.ok(retVal).build();
   }
 
-  private DruidDataSource getDataSource(final String dataSourceName)
+  private ImmutableDruidDataSource getDataSource(final String dataSourceName)
```
Comment (leventov) [inline server/src/main/java/io/druid/server/http/DatasourcesResource.java]: Annotate `@Nullable`

File: server/src/main/java/io/druid/server/http/DatasourcesResource.java [hunk 16]
```diff
@@ -496,40 +493,32 @@ public Response getSegmentDataSourceTiers(
...
-          segmentMap.put(segment.getIdentifier(), segment);
-        }
+    for (ImmutableDruidDataSource dataSource : dataSources) {
+      Iterable<DataSegment> segments = dataSource.getSegments();
+      for (DataSegment segment : segments) {
+        segmentMap.put(segment.getIdentifier(), segment);
       }
     }
 
     return new DruidDataSource(
```
Comment (leventov) [inline server/src/main/java/io/druid/server/http/DatasourcesResource.java]: This is a strange expression, why ImmutableDruidDataSource couldn't be constructed directly?

File: server/src/main/java/io/druid/server/http/DatasourcesResource.java [hunk 17]
```diff
@@ -496,40 +493,32 @@ public Response getSegmentDataSourceTiers(
...
-        );
-
-    List<DruidDataSource> validDataSources = Lists.newArrayList();
-    for (DruidDataSource dataSource : dataSources) {
-      if (dataSource != null) {
-        validDataSources.add(dataSource);
-      }
-    }
-    if (validDataSources.isEmpty()) {
+    List<ImmutableDruidDataSource> dataSources = serverInventoryView
```
Comment (leventov) [inline server/src/main/java/io/druid/server/http/DatasourcesResource.java]: Why rename "validDataSources" to simply "dataSources"?

File: server/src/main/java/io/druid/server/http/InventoryViewUtils.java [hunk 18]
```diff
@@ -37,42 +33,23 @@
...
-                )
-            )
-        )
-    );
-    return dataSources;
+    return serverInventoryView.getInventory()
+                              .stream()
+                              .flatMap(server -> server.getDataSources().stream())
+                              .map(DruidDataSource::toImmutableDruidDataSource)
+                              .collect(
```
Comment (leventov) [inline server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: Why not just `toSet()`?

File: server/src/test/java/io/druid/metadata/MetadataSegmentManagerTest.java [hunk 19]
```diff
@@ -123,8 +124,8 @@ public void testPoll()
         manager.getAllDatasourceNames()
     );
     Assert.assertEquals(
-        ImmutableSet.of(segment1, segment2),
-        manager.getInventoryValue("wikipedia").getSegments()
+        Sets.newHashSet(segment1, segment2),
```
Comment (leventov) [inline server/src/test/java/io/druid/metadata/MetadataSegmentManagerTest.java]: Why this change?

File: server/src/main/java/io/druid/client/ImmutableDruidDataSource.java (hunk 5 above)
Comment (drcrallen) [reply server/src/main/java/io/druid/client/ImmutableDruidDataSource.java]: Probably, my bad

File: server/src/main/java/io/druid/client/DruidDataSource.java (hunk 9 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/DruidDataSource.java]: Done.

File: server/src/main/java/io/druid/client/DruidDataSource.java (hunk 10 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/DruidDataSource.java]: Removed.

File: server/src/main/java/io/druid/client/ImmutableDruidDataSource.java (hunk 11 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/ImmutableDruidDataSource.java]: Thanks. Removed.

File: server/src/main/java/io/druid/client/ImmutableDruidServer.java (hunk 12 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/ImmutableDruidServer.java]: I think it's not nullable because ```DruidServer``` doesn't check in ```compareTo()``` and ```ImmutableDruidServer``` is created from ```DruidServer```. Added null check in the constructor of ```DruidServerMetadata```.

Also changed to use ```DruidServerMetadata``` in ```equals()``` and ```hashCode()``` as in https://github.com/druid-io/druid/pull/5056.

File: server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java (hunk 13 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java]: Done.

File: server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java (hunk 14 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java]: Done.

File: server/src/main/java/io/druid/server/http/DatasourcesResource.java (hunk 15 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/DatasourcesResource.java]: Done.

File: server/src/main/java/io/druid/server/http/DatasourcesResource.java (hunk 17 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/DatasourcesResource.java]: I don't understand what "valid" means here. In the original code, ```dataSources``` may contain null dataSources, but it means that some servers don't contain the dataSource of the given name, not the dataSource is invalid.

File: server/src/main/java/io/druid/server/http/DatasourcesResource.java (hunk 16 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/DatasourcesResource.java]: Ah, it was somewhat complicated before, but now it's fine. Changed to directly construct ImmutableDruidDataSource.
Note that ImmutableMap.Builder cannot be used for segmentMap because the same segment identifier can be inserted multiple times. I'm not sure this is the valid behavior and want to keep it as it is.

File: server/src/main/java/io/druid/server/http/InventoryViewUtils.java (hunk 18 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: Is it possible to collect as a TreeSet? This originally returns a TreeSet, and I don't want to change it.

File: server/src/test/java/io/druid/metadata/MetadataSegmentManagerTest.java (hunk 19 above)
Comment (jihoonson) [reply server/src/test/java/io/druid/metadata/MetadataSegmentManagerTest.java]: Reverted.

File: server/src/main/java/io/druid/server/http/InventoryViewUtils.java (hunk 18 above)
Comment (leventov) [reply server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: Could you please check this question, and leave a comment in code? as the method return type is just `Set`, not `NavigableSet` or `SortedSet`, it's suspicious

Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: Ah, I just realized this method is always called by the below ```getSecuredDataSources()``` and it converts the TreeSet to an ImmutableSet anyway. Changed to ```toSet()```.

File: server/src/main/java/io/druid/client/ImmutableDruidDataSource.java [hunk 20]
```diff
@@ -73,7 +74,6 @@ public DataSegment getSegment(String segmentIdentifier)
   @Override
   public String toString()
   {
-    // idToSegments is intentionally ignored because it is usually large
```
Comment (leventov) [inline server/src/main/java/io/druid/client/ImmutableDruidDataSource.java]: But maybe rather need to keep ignore it?) If it used to be ignored for a good reason

File: server/src/main/java/io/druid/server/http/InventoryViewUtils.java (hunk 18 above)
Comment (leventov) [reply server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: On ImmutableDruidDataSource, `idToSegments` hash code computation may take long

Comment (jihoonson) [reply server/src/main/java/io/druid/server/http/InventoryViewUtils.java]: Hmm, yes. In addition, a unit test expects the result is sorted by the name. I reverted this and changed the return type to SortedSet.

File: server/src/main/java/io/druid/client/ImmutableDruidDataSource.java (hunk 20 above)
Comment (jihoonson) [reply server/src/main/java/io/druid/client/ImmutableDruidDataSource.java]: Good point. Added the number of segments instead of detailed segments.
